### PR TITLE
[MRG] Simple version Image Pipeline Implemented

### PR DESCRIPTION
**Pull request by jhfjhfj1**
### Pull Request Status 
MRG

### Details of the Pull Request
Implemented a simple version of ImageBlock and used it for the ImageClassifier and ImageRegressor.

### CODE REVIEW

File: autokeras/auto/image.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        new_space = hyperparameters.space[:]
+        if not self.allow_new_parameters and set(old_space) != set(new_space):
+            diff = set(new_space) - set(old_space)
+            raise RuntimeError(
+                'The hypermodel has requested a parameter that was not part '
+                'of `hyperparameters`, '
+                'yet `allow_new_parameters` is set to False. '
+                'The unknown parameters are: {diff}'.format(diff=diff))
+
+        datagen = tf.keras.preprocessing.image.ImageDataGenerator(
```
Comment (fchollet) [inline autokeras/auto/image.py]: Prefer using tf.data for a pipeline. Reference: https://www.tensorflow.org/tutorials/load_data/images

You'll probably need some basic utility for data augmentation with tf.data, let's touch base on that later.

File: autokeras/auto/image.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+from autokeras import tuner
+
+
+class ImageTuner(tuner.SequentialRandomSearch):
+
+    def _run(self, hyperparameters, fit_kwargs):
+        # Build a model instance.
+        model = self.hypermodel.build(hyperparameters)
+
+        # Optionally disallow hyperparameters defined on the fly.
```
Comment (fchollet) [inline autokeras/auto/image.py]: This block should be abstracted as a method on the base class.

File: autokeras/auto/image.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+            # randomly flip images
+            horizontal_flip=True,
+            # randomly flip images
+            vertical_flip=False)
+
+        datagen.fit(fit_kwargs['x'])
+        fit_kwargs['batch_size'] = min(len(fit_kwargs['x']),
+                                       fit_kwargs.get(
+                                           'batch_size',
+                                           const.Constant.BATCH_SIZE))
```
Comment (fchollet) [inline autokeras/auto/image.py]: The batch size should be at least configurable (e.g. in `__init__`)

File: autokeras/auto/image.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        return model, feedback
+
+
+class ImageSupervised(auto_pipeline.AutoPipeline):
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.image_block = hyper_block.ImageBlock()
+        self.head = None
+        self.x_normalizer = processor.Normalizer()
```
Comment (fchollet) [inline autokeras/auto/image.py]: just `normalizer` or `input_normalizer`, in general avoid single-letter notations

File: autokeras/auto/image.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+    def fit(self, x=None, y=None, **kwargs):
+        self.tuner = ImageTuner(self, metrics=self.head.metrics)
+        self.x_normalizer.fit(x)
+        self.inputs = [hyper_node.ImageInput()]
+        super().fit(x=self.x_normalizer.transform(x), y=y, **kwargs)
```
Comment (fchollet) [inline autokeras/auto/image.py]: In the future `normalizer` will be a preprocessing layer part of the model (added in `build`).

File: autokeras/auto/image.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+        # Train model
+        history = model.fit_generator(data_flow, **temp_fit_kwargs)
+
+        metric_name = model.metrics_names[1]
+        feedback = history.history['val_' + metric_name][-1]
+        return model, feedback
+
+
+class ImageSupervised(auto_pipeline.AutoPipeline):
```
Comment (fchollet) [inline autokeras/auto/image.py]: This is a base class, right? Perhaps `SupervisedImagePipeline` would be a more explicit name.

File: autokeras/auto/processor.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
+import numpy as np
+
+
+class OneHotEncoder:
```
Comment (fchollet) [inline autokeras/auto/processor.py]: `(object)`

File: autokeras/auto/processor.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    """A class that can format data.
+
+    This class provides ways to transform data's classification label into
+    vector.
+
+    Attributes:
```
Comment (fchollet) [inline autokeras/auto/processor.py]: Use markdown conventions:

```
# Attributes
```

File: autokeras/auto/processor.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+          n_classes: The number of classes in the classification problem.
+          labels: The number of labels.
+          label_to_vec: Mapping from label to vector.
+          int_to_label: Mapping from int to label.
+    """
+
+    def __init__(self):
+        """Initialize a OneHotEncoder"""
+        self.data = None
+        self.n_classes = 0
```
Comment (fchollet) [inline autokeras/auto/processor.py]: `num_`

File: autokeras/auto/processor.py (hunk 7 above)
Comment (fchollet) [reply autokeras/auto/processor.py]: Keep in mind that this whole file will end up deprecated with preprocessing layers and tf.data. Don't put too much work into it.

File: autokeras/auto/processor.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    """
+
+    def __init__(self):
+        self.max_val = None
+        self.mean = None
+        self.std = None
+
+    def fit(self, data):
+        self.max_val = data.max()
+        data = data / self.max_val
```
Comment (fchollet) [inline autokeras/auto/processor.py]: Prefer following the API of the Normalization layer (mean and stddev normalization) to avoid issues when converting to a preprocessing layer.

File: autokeras/auto/image.py (hunk 3 above)
Comment (jhfjhfj1) [reply autokeras/auto/image.py]: This is already configurable from fit function of AutoModel. fit_kwargs is the **kwargs of fit function of AutoModel.

File: autokeras/auto/image.py (hunk 1 above)
Comment (jhfjhfj1) [reply autokeras/auto/image.py]: I will submit another pull request for changing this one.

**Comment by jhfjhfj1** [comment]
The comments are all addressed except for using tf.data. It is ready for another review. Thanks. @fchollet

File: autokeras/auto/auto_model.py [hunk 11]
```diff
@@ -96,21 +84,29 @@ def __init__(self,
         self._hypermodels = []
         self._hypermodel_to_id = {}
         self._build_network()
+        self.tuner = tuner.SequentialRandomSearch(
+            self,
+            objective=self._get_metrics())
 
-    def build(self, hp):
+    def build(self, hp, **kwargs):
```
Comment (fchollet) [inline autokeras/auto/auto_model.py]: Where are `**kwargs` used?

If another piece of code need to programatically call `build`, how can it know what arguments to pass?

File: autokeras/auto/auto_model.py [hunk 12]
```diff
@@ -146,20 +143,24 @@ def _build_network(self):
...
+                # Check at least one output node of the hypermodel is in the
+                # interested nodes.
+                if not any([output_node in self._node_to_id for output_node in
+                            hypermodel.outputs]):
                     continue
                 self._add_hypermodel(hypermodel)
                 for output_node in hypermodel.outputs:
                     # The node is not visited and in interested nodes.
-                    if output_node not in visited_nodes and output_node in self._node_to_id:
+                    if output_node not in visited_nodes \
```
Comment (fchollet) [inline autokeras/auto/auto_model.py]: Prefer

```
if (output_node not in visited_nodes and
     output_node in self._node_to_id):
```

File: autokeras/auto/auto_model.py (hunk 11 above)
Comment (jhfjhfj1) [reply autokeras/auto/auto_model.py]: This is removed since we are not using the **kwargs in the build function.
